python/tk_desktop/rpc.py

```diff
@@ -268,3 +393,540 @@ def close(self):
+    """
+    Return the right server type based on the pipe.
+    """
+    if pipe.startswith("http://"):
+        return HttpRPCServerThread
+    else:
+        return MultiprocessingRPCServerThread
+
+
+# These are present for backwards compatibility. Clients can take over part of the
+# bootstrap code and they'll likely to want to instantiate those classes.
+# When they do, we want them to use the Http based ones since those are compatible
+# with Python 2 and 3.
+RPCProxy = HttpRPCProxy
+RPCServerThread = HttpRPCServerThread
+
+
+class DualRPCServer(object):
+    """
+    Run both the Http and Multiprocessing based servers at the same time.
+
+    This allows older and newer versions of tk-desktop to be launched in
+    the without having to try to guess which server to spin up.
+
+    Such a guess would be impossible anyway because there is no way to
+    guarantee which version of tk-desktop will be launched in the background.
+
+    This class mimicks the interface of MultiprocessingRPCServerThread and
+    HttpRPCServerThread so they are interchangeable in the code. Since both
+    of these classes were threads, they also implement thread methods like
+    start, is_alive and join.
+    """
+
+    def __init__(self, engine):
+        self._authkey = str(uuid.uuid1()).encode()
+        self._multiprocessing_thread = MultiprocessingRPCServerThread(
+            engine, self._authkey
+        )
+        self._http_thread = HttpRPCServerThread(engine, self._authkey)
+
+    @property
+    def pipes(self):
+        """
+        Connection strings to both servers.
+
+        :returns: Tuple of connection strings.
+        """
+        return (self._multiprocessing_thread.pipe, self._http_thread.pipe)
+
+    @property
+    def engine(self):
+        """
+        Current engine.
+        """
+        return self._multiprocessing_thread.engine
+
+    @property
+    def authkey(self):
+        """
+        Credential to connect to this server.
+        """
+        return self._authkey
+
+    def is_closed(self):
+        """
+        :returns: ``True`` if the server is closed, ``False`` if not.
+        """
+        # Both are closed or both are opened, so no need to ask both.
+        return self._multiprocessing_thread.is_closed()
+
+    def list_functions(self):
+        """
+        Return the list of functions registered with the server.
+        """
+        # Both have the same functions, so no need to ask both.
+        return self._multiprocessing_thread.list_functions()
+
+    def register_function(self, func, name=None):
+        """
+        Register a function with the server.
+
+        :param callable func: Method to call.
+        :param name: Optional. If set, the function will be named as the parameter states.
+            Otherwise, func.__name__ will be used.
+        """
+        self._multiprocessing_thread.register_function(func, name)
+        self._http_thread.register_function(func, name)
```

Comment:
I'm sure this is fine, but since this code is handling both multiprocessing and http servers, do we increase the risk that if one of them errors for some reason, everything will stop even if we weren't using that type of server?

Reply:
If this breaks, then I've written some really terrible tests. 😜 

Besides, if registering a function fails, then the Desktop will crash and give a pop-up.